File: Top 100 Codes/04-Important Codes related to Arrays/30-Juggling algorithm for array rotation/Jugglingalgorithmforarrayrotation.cpp
#include <iostream>
#include <algorithm> // For std::gcd (C++17 and later)
using namespace std;

// Function to rotate array using Juggling Algorithm
void juggleRotate(int arr[], int n, int d) {
    d = d % n; // Handle cases where d >= n
    int gcd = __gcd(n, d); // Find GCD of n and d

    for (int i = 0; i < gcd; i++) {
        int temp = arr[i]; // Store first element of the set
        int j = i;

        while (true) {
            int k = j + d;
            if (k >= n) 
               k -= n; // Wrap around if out of bounds

            if (k == i) 
               break; // If cycle is complete

            arr[j] = arr[k]; // Move next element to current position
            j = k;
        }

        arr[j] = temp; // Place stored value at the correct position
    }
}

// Function to print an array
void Display(int arr[], int n) {
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

// Driver Code
int main() {
    int arr[] = {1, 2, 3, 4, 5, 6, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    int d = 3; // Rotate left by 3 positions

    cout << "Original Array: ";
    Display(arr, n);

    juggleRotate(arr, n, d);

    cout << "Rotated Array: ";
    Display(arr, n);

    return 0;
}

File: Top 100 Codes/04-Important Codes related to Arrays/30-Juggling algorithm for array rotation/Jugglingalgorithmforarrayrotation.py
import math

def juggle_rotate(arr, d, n):
    d = d % n  # Handle cases where d >= n
    gcd = math.gcd(n, d)  # Find GCD of n and d

    for i in range(gcd):
        temp = arr[i]  # Store the first element of the set
        j = i

        while True:
            k = j + d
            if k >= n:
                k -= n  # Wrap around if out of bounds

            if k == i:
                break  # If cycle is complete

            arr[j] = arr[k]  # Move next element to current position
            j = k

        arr[j] = temp  # Place stored value at the correct position

def print_array(arr):
    print(" ".join(map(str, arr)))

# Driver code
arr = [1, 2, 3, 4, 5, 6, 7]
d = 3  # Rotate left by 3 positions
n = len(arr)

print("Original Array:")
print_array(arr)

juggle_rotate(arr, d, n)

print("Rotated Array:")
print_array(arr)
